tower-log/src/lib.rs

```diff
@@ -0,0 +1,452 @@
+    };
+    (target: $target:expr, $($rest:tt)+) => {
+        log_errors!($($rest),+).target($target)
+    };
+    (level: $level:expr, $inner:expr) => {
+        $crate::LogErrors::new($inner)
+            .in_module(module_path!())
+            .at_location(file!(), line!())
+            .at_level($level)
+    };
+}
+
+/// Wrap a `Service` or `NewService` with `LogResponses` middleware.
+///
+/// Unlike using `LogResponses::new`, this macro will configure the returned
+/// middleware to log messages with the module path and file/line location of
+/// the _call site_, as using the `log!` macro in that file would.
+#[macro_export]
+macro_rules! log_responses {
+    ($inner:expr) => {
+        log_responses!(level: ::log::Level::Debug, $inner)
+    };
+    (target: $target:expr, $($rest:tt)+) => {
+        log_responses!($($rest),+).target($target)
+    };
+    (level: $level:expr, $inner:expr) => {
+        $crate::LogResponses::new($inner)
+            .in_module(module_path!())
+            .at_location(file!(), line!())
+            .at_level($level)
+    };
+}
+/// Logs error responses.
+#[derive(Clone, Debug)]
+pub struct LogErrors<T> {
+    inner: T,
+    level: log::Level,
+    target: Option<&'static str>,
+    module_path: Option<&'static str>,
+    file: Option<&'static str>,
+    line: Option<u32>,
+}
+
+/// Logs successful responses.
+#[derive(Clone, Debug)]
+pub struct LogResponses<T> {
+    inner: T,
+    level: log::Level,
+    not_ready: bool,
+    target: Option<&'static str>,
+    module_path: Option<&'static str>,
+    file: Option<&'static str>,
+    line: Option<u32>,
+}
+
+
+// ===== impl LogErrors =====
+
+impl<T> LogErrors<T> {
+
+    /// Construct a new `LogErrors` middleware that wraps the given `Service`
+    /// or `NewService`.
+    ///
+    /// The log level will default to `Level::Error` but may be changed with
+    /// the [`at_level`] function.
+    ///
+    /// # Note
+    ///
+    /// Unless the module path of the `LogErrors` middleware is changed with
+    /// the [`in_module`]  methods, log records produced by the returned
+    /// middleware will always have the module path `tower-log-errors`. It may
+    /// be preferred to use the [`log_errors!`] macro instead, as it will
+    /// produce log records which appear to have been produced at the call site.
+    ///
+    /// [`at_level`]: struct.LogErrors.html#method.at_level
+    /// [`in_module`]: struct.LogErrors.html#method.in_module
+    /// [`log_errors!`]: macro.log_errors.html
+    pub fn new(inner: T) -> Self {
+        LogErrors {
+            inner,
+            level: log::Level::Error,
+            target: None,
+            module_path: None,
+            file: None,
+            line: None,
+        }
+    }
+
+    /// Set the log level of the produced log records.
+    ///
+    /// Log records will be logged at the `Error` level by default.
+    pub fn at_level(mut self, level: log::Level) -> Self {
```

Comment:
Could these be broken out into a `Builder`. Then we can match the fn names from [`RecordBuilder`](https://docs.rs/log/0.4.1/log/struct.RecordBuilder.html):

* `level`
* `target`
* `file`
* `line`

etc...